utils/logalloc.cc

```diff
@@ -693,6 +693,99 @@ struct segment_descriptor : public log_heap_hook<segment_descriptor_hist_options
 
 using segment_descriptor_hist = log_heap<segment_descriptor, segment_descriptor_hist_options>;
 
+class segment_store_backend {
+protected:
+    memory::memory_layout _layout;
+    bool _overlaps_with_non_lsa_memory;
+    uintptr_t _segments_base; // The address of the first segment
+
+public:
+    explicit segment_store_backend(memory::memory_layout layout, bool overlaps_with_non_lsa_memory)
+        : _layout(layout)
+        , _overlaps_with_non_lsa_memory(overlaps_with_non_lsa_memory)
+        , _segments_base(align_up(_layout.start, static_cast<uintptr_t>(segment::size)))
+    { }
+    memory::memory_layout memory_layout() const noexcept { return _layout; }
+    uintptr_t segments_base() const noexcept { return _segments_base; }
+    bool overlaps_with_non_lsa_memory() const noexcept { return _overlaps_with_non_lsa_memory; }
+    virtual void* alloc_segment_memory() noexcept = 0;
+    virtual void free_segment_memory(void* seg) noexcept = 0;
+    virtual size_t free_memory() const noexcept = 0;
+};
+
+class seastar_memory_segment_store_backend : public segment_store_backend {
+public:
+    seastar_memory_segment_store_backend()
+        : segment_store_backend(memory::get_memory_layout(), true)
+    { }
+    virtual void* alloc_segment_memory() noexcept override {
+        return aligned_alloc(segment::size, segment::size);
+    }
+    virtual void free_segment_memory(void* seg) noexcept override {
+        ::free(seg);
+    }
+    virtual size_t free_memory() const noexcept override {
+        return memory::stats().free_memory();
+    }
+};
+
+class mmap_memory_segment_store_backend : public segment_store_backend {
```

Comment:
I don't understand why we need a mmap backend. The debug allocator (below) uses aligned_alloc.

As far as I can tell, the critical different is using memory layout for the  segment address -> descriptor translation, vs. using a map.

Reply:
I will change this to go through the allocator. Indeed there is not much point in using mmap.

Reply:
The problem with the current state of art is that the the two stores use completely different method to map segment address -> descriptor translation. If I wanted to fit both into a common interface I'd have to virtualize the segment address to descriptor mapping, which is quite commonly used. To avoid that I standardize  on the memory layout based mapping and just virtualize how the backing memory is obtained and managed, as this is used much more rarely.

Reply:
Ah, so mmap is used obtain all the segments, not an individual segment?

I'm confused. We already have two different segment -> descriptor mappings (for release and debug mode) and two different ways to obtain the memory. You're adding a third way to obtain memory, but using the division method for segment->descriptor?

It's probably the right thing if translations are frequent.

Reply:
> Ah, so mmap is used obtain all the segments, not an individual segment?

Yes. I wanted to emulate the seastar allocator which owns a large contiguous area of memory, allocating some of it to segments. This allows for simple segment -> descriptor mapping, via calculating the segment's offset in the layout and using that as an index. I wanted to emulate this in debug mode, obtaining a large continuous area of memory, dedicated to segments, so we can use a single mapping logic across release/debug modes.
I used mmap to obtain this memory, but getting it from the standard allocator works just as well.

After this patch we have two methods for obtaining the memory area that will be used to back segments, and a single method for mapping segments to descriptors.

Reply:
It's not good for debug mode, since if we overrun a segment, we hit another segment without noticing. One of asan's benefits is to prevent overruns and we lose it with the release mode segment allocation.

Reply:
We can leave gaps between, although this will increase waste (gaps would have to be segment sized, as segments are aligned to `segment::size`). I wonder if poisoning forces the kernel to back poisoned pages with real memory. Otherwise, the waste might remain theoretical only.

Or we could just keep the current debug-mode segment store for debug-only and use the new one for tools only. I'd like to avoid that if possible.

Reply:
No, poisoning wastes the memory. asan uses separate mmaps to large trap out-of-bounds violations and poisoning to trap small violations and uninitialized variable use.

Why not keep the debug-mode segment store? It works, and messing with it is liable to reduce our ability to catch bugs (since we don't test it).

Reply:
Ok, I will keep the current debug store.